### ISSUE TITLE
Allow apps packaged as native Linux system packages to use a non-default Python

### ISSUE THREAD

**Issue by rmartin16** [beeware/briefcase#1976]
### What is the problem or limitation you are having?

Currently, Briefcase enforces that apps packaged as native Linux system packages target the default Python for the distro. For instance, Ubuntu 22.04 symlinks Python 3.10 from `/usr/bin/python3`. However, most distros now allow side-by-side Python installations. So, on Ubuntu 22.04, Python3.11 can be installed; however, Briefcase's native Linux system packages cannot target non-default Python versions.

### Describe the solution you'd like

Add support to target non-default Python versions when packaging native Linux system packages.

### Describe alternatives you've considered

Status quo.

**Comment by freakboy3742** [comment]
Agreed this would be desirable, provided the upstream provides an easy way to get these "non default" Pythons as part of system packaging. On Ubuntu, that means you need to have a Deadsnakes PPA available - but you can't specify installing a PPA as a dependency of a package (as far as I know), so an app packaged based on a Deadsnakes Python is *usable*, but requires additional work on the part of the user. I'm not as familiar with the situation on Fedora/Redhat.

That said, I guess we could make this a buildtime warning - if the system *provides* a non-default Python version, allow it to be used; but warn that the resulting package won't necessarily be as immediately usable, so they're going to need to educate users on how to add a PPA (or whatever).

**Comment by rmartin16** [comment]
I was partly prompted to create this from this Python Discourse [discussion](https://discuss.python.org/t/publish-linux-installer-on-python-org/53760/81). Detailed there is the current state of Python availability by Linux distributors.

In short, some distros (like Fedora) provide access to all non-EOL Pythons; but all distros provide access to, at least, two Pythons. For instance, Ubuntu 22.04 provides Python 3.10 and 3.11 without user-added PPAs.
```
❯ docker run --rm -it ubuntu:22.04 bash -c 'apt update && apt install python3 python3.11'
Unable to find image 'ubuntu:22.04' locally
22.04: Pulling from library/ubuntu
857cc8cb19c0: Pull complete 
Digest: sha256:adbb90115a21969d2fe6fa7f9af4253e16d45f8d4c1e930182610c4731962658
Status: Downloaded newer image for ubuntu:22.04
Get:1 http://security.ubuntu.com/ubuntu jammy-security InRelease [129 kB]
[...]
Get:18 http://archive.ubuntu.com/ubuntu jammy-backports/universe amd64 Packages [33.7 kB]
Fetched 33.8 MB in 2s (16.1 MB/s)                            
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
5 packages can be upgraded. Run 'apt list --upgradable' to see them.
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
The following additional packages will be installed:
  libexpat1 libmpdec3 libpython3-stdlib libpython3.10-minimal libpython3.10-stdlib libpython3.11-minimal libpython3.11-stdlib libreadline8 libsqlite3-0 media-types python3-minimal python3.10
  python3.10-minimal python3.11-minimal readline-common
Suggested packages:
  python3-doc python3-tk python3-venv python3.10-venv python3.10-doc binutils binfmt-support python3.11-venv python3.11-doc readline-doc
The following NEW packages will be installed:
  libexpat1 libmpdec3 libpython3-stdlib libpython3.10-minimal libpython3.10-stdlib libpython3.11-minimal libpython3.11-stdlib libreadline8 libsqlite3-0 media-types python3 python3-minimal
  python3.10 python3.10-minimal python3.11 python3.11-minimal readline-common
0 upgraded, 17 newly installed, 0 to remove and 5 not upgraded.
Need to get 12.1 MB of archives.
After this operation, 44.9 MB of additional disk space will be used.
Do you want to continue? [Y/n] 
```

**Comment by freakboy3742** [comment]
Oof... that's a pretty damning analysis of the Debian-alikes...

I agree it makes sense to be able to use any Python that is available as a system package. On Debian/Ubuntu, we might want to issue a warning if we detect the Python version in use has come from Deadsnakes. We probably also want to retain the checks to make sure that the installed version is in the "system" folder, not a local source/pyenv/Conda install; but otherwise, I don't see any reason we should restrict the version to just "what is the default python3" on a platform.

**Comment by rmartin16** [comment]
> On Debian/Ubuntu, we might want to issue a warning if we detect the Python version in use has come from Deadsnakes. We probably also want to retain the checks to make sure that the installed version is in the "system" folder, not a local source/pyenv/Conda install

Coming back to this....currently, Briefcase does not care if the Python being used to run Briefcase is not a Python installed by the system itself; Briefcase only verifies that the version of Python running Briefcase is the same as the version of the default system Python. I believe this status quo should remain. I am not sure I see benefits from requiring users to use a system Python to run Briefcase.

**Comment by freakboy3742** [comment]
Broadly speaking, I agree that we should be as liberal as possible in accepting Python versions. 

As for reasons why me *might* restrict - I seem to recall there are some complications if you use a conda-installed or pyenv-installed version of Python instead of the system package. However, it's been a while since I've tried, so my memory might be failing.

**Comment by freakboy3742** [comment]
@meichthys It's not clear to me what you're reporting here. The error message you're receiving is accurate - the python you're using isn't the system Python. At present, that's the correct behavior; changing this behavior is what this ticket is proposing. 

That said, `pipenv` is unlikely to be a configuration that Briefcase supports - we need to be able to specify a system package that provides the Python interpreter, and pipenv doesn't provide that. The only way to make this work would be to bundle a python executable as part of the package - and that is the sort of thing that makes Linux distro maintainers *very* grumpy.

**Comment by freakboy3742** [comment]
A follow up on this - with the release of Briefcase 0.4.0, we have some additional options here.

Briefcase 0.4.0 has the ability to create and use virtual environments; so, as long as the virtual environment used to build the final system app is created using the system python (or *a* system provided python - something that can be added to a DEB/RPM requirements list), the rest of the development process *should* be compatible with any Python executable.

**Comment by abdnh** [comment]
Would embedding a Python build from https://github.com/astral-sh/python-build-standalone solve the problem and make Linux builds much more portable?

Happy to research this further and possibly work on a PR if you're open to the idea.

For context, I decided to go for a hybrid Briefcase/PyInstaller solution for [Anki](https://github.com/ankitects/anki) for better portability across different distros: https://github.com/ankitects/anki/issues/4622

**Comment by abdnh** [comment]
Oh, I just found out that AppImage/Flatpak already uses python-build-standalone.

>The only way to make this work would be to bundle a python executable as part of the package - and that is the sort of thing that makes Linux distro maintainers very grumpy.

If providing that for system packages is not a good option, what about introducing a zip format? For Anki, that'd actually be our preferred format so we don't have to worry about building system packages. Currently we distribute a zip with some custom install scripts.

Let me know if that should be discussed in a separate issue.

**Comment by freakboy3742** [comment]
> If providing that for system packages is not a good option, what about introducing a zip format? For Anki, that'd actually be our preferred format so we don't have to worry about building system packages. Currently we distribute a zip with some custom install scripts.
> 
> Let me know if that should be discussed in a separate issue.

A separate packaging format should definitely be a separate issue; it should probably start as a discussion first; see #2817.
